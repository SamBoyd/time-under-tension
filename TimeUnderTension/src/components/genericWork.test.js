import React from 'react';
import configureStore from 'redux-mock-store';
import {Provider} from 'react-redux';
import renderer from "react-test-renderer";
import {v4 as uuidv4} from "uuid";
import '@testing-library/jest-dom/extend-expect'

import GenericWork from "./genericWork";

jest.useFakeTimers()


const mockStore = configureStore([]);

describe('GenericWork component', () => {
    let store;
    let component;

    beforeEach(() => {
        store = mockStore({
            settings: {
                defaultRestTime: 10,
                defaultWorkTimeStart: 20,
                defaultWorkTimeEnd: 30,
            },
            workoutState: {
                name: 'Workout',
                id: uuidv4(),
                created_at: (new Date()).toISOString(),
                work: ['work_1'],
                started_at: null,
                finished_at: null,
            },
            work: [
                {
                    id: 'work_1',
                    created_at: (new Date()).toISOString(),
                    exercise: {name: 'ex1', category: 'cat1'},
                    sets: ['set1'],
                    restTime: null,
                    workTimeStart: null,
                    workTimeEnd: null,
                    finished: false,
                }
            ],
            set: [
                {
                    id: 'set1',
                    numberReps: 12,
                    weight: 40,
                    workTime: null,
                    finished: false,
                }
            ],
            timer: {},
        });
    });

    const getComponent = (
        id = 'work1',
        exercise = {name: 'ex1', category: 'cat1'},
        sets = ['set1'],
        restTime = 10,
        workTimeStart = 20,
        workTimeEnd = 30,
        finished = null,
        active = true,
        workIndex = 0,
        fireAddSet = jest.fn(),
        fireRemoveSet = jest.fn(),
        fireChangeRestTime = jest.fn(),
        fireChangeWorkTimeStart = jest.fn(),
        fireChangeWorkTimeEnd = jest.fn(),
        removeWork = jest.fn(),
        moveWorkUp = jest.fn(),
        moveWorkDown = jest.fn(),
        displayChangeActiveWork = jest.fn(),
        previousSets = [],
    ) => {
        return renderer.create(
            <Provider store={store}>
                <GenericWork
                    id={id}
                    exercise={exercise}
                    sets={sets}
                    restTime={restTime}
                    workTimeStart={workTimeStart}
                    workTimeEnd={workTimeEnd}
                    finished={finished}
                    active={active}
                    workIndex={workIndex}
                    fireAddSet={fireAddSet}
                    fireRemoveSet={fireRemoveSet}
                    fireChangeRestTime={fireChangeRestTime}
                    fireChangeWorkTimeStart={fireChangeWorkTimeStart}
                    fireChangeWorkTimeEnd={fireChangeWorkTimeEnd}
                    removeWork={removeWork}
                    moveWorkUp={moveWorkUp}
                    moveWorkDown={moveWorkDown}
                    displayChangeActiveWork={displayChangeActiveWork}
                    previousSets={previousSets}
                />
            </Provider>
        )
    }

    it('should render with given state from Redux store', () => {
        const component = getComponent()
        expect(component.toJSON()).toMatchSnapshot();
    });

    xit('displays zero values for the rest and work time', () => {
        const component = getComponent(
            'work1',
            {name: 'ex1', category: 'cat1'},
            ['set1'],
            0,
            0,
            0,
            null,
            true,
            0,
            jest.fn(),
            jest.fn(),
            jest.fn(),
            jest.fn(),
            jest.fn(),
            jest.fn(),
            jest.fn(),
            jest.fn(),
            jest.fn(),
            [],
        )

        expect(component.root.findByProps({testId: 'workTime_work1'}).textContent).toBe('Target tension for 0 to 0 sec')
        expect(component.root.findByProps({testId: 'restTime_work1'}).textContent).toBe('Rest for 0 sec')
    })

    xit('can edit the work', () => {
        renderer.act(() => {
            component.root.findByProps({testId: 'toggleEditWork'}).props.onPress()
        })

        renderer.act(() => {
            component.root.findByProps({testId: 'saveEditWork'}).props.onPress()
        })


        expect(component.root.children[0].props.fireChangeRestTime).not.toBeCalled()
        expect(component.root.children[0].props.fireChangeWorkTimeStart).not.toBeCalled()
        expect(component.root.children[0].props.fireChangeWorkTimeEnd).not.toBeCalled()
        renderer.act(() => {
            component.root.findByProps({testId: 'toggleEditWork'}).props.onPress()
        })
        renderer.act(() => {
            component.root.findByProps({testId: 'optionRestTime_1'}).click()
        })
        renderer.act(() => {
            component.root.findByProps({testId: 'saveEditWork'}).props.onPress()

        })

        expect(component.root.findByProps({testId: 'restTime_work1'})).toHaveTextContent('Rest time for 2')
    })
})
