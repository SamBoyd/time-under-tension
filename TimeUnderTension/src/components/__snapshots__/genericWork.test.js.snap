// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenericWork component should render with given state from Redux store 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#262521",
      "borderRadius": 8,
      "borderWidth": 0,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "baseline",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#CECABA",
          "fontSize": 18,
          "fontWeight": "bold",
          "marginBottom": 15,
          "textAlign": "center",
        }
      }
      testID="cardTitle"
    >
      ex1
    </Text>
    <View
      style={
        Object {
          "flexDirection": "row",
          "textAlignVertical": "center",
        }
      }
    >
      <View
        style={
          Object {
            "overflow": "hidden",
          }
        }
      >
        <View
          accessibilityState={
            Object {
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "justifyContent": "center",
              }
            }
          >
            <Text
              allowFontScaling={false}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#CECABA",
                    "fontSize": 30,
                  },
                  Object {
                    "backgroundColor": "transparent",
                  },
                  Object {
                    "fontFamily": "Material Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
              testID="iconIcon"
            >
              
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    replaceTheme={[Function]}
    style={
      Object {
        "borderBottomColor": "#bcbbc1",
        "borderBottomWidth": 0.5,
        "marginBottom": 15,
      }
    }
    updateTheme={[Function]}
  />
  <View
    style={
      Object {
        "flexDirection": "row",
        "textAlign": "center",
        "textAlignVertical": "center",
      }
    }
    viewStyle={
      Object {
        "textAlign": "center",
        "textAlignVertical": "center",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#CECABA",
          "fontSize": 16,
        }
      }
      testId="workTime_work1"
    >
      Target tension for 
      20
       to 
      30
       seconds 
    </Text>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "textAlignVertical": "center",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#CECABA",
          "fontSize": 16,
        }
      }
      testId="restTime_work1"
    >
      Rest time for 
      10
       seconds
    </Text>
  </View>
  <View
    style={
      Object {
        "marginTop": 0,
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
      <View
        style={
          Object {
            "alignItems": "baseline",
            "borderBottomWidth": 1,
            "borderColor": "#CDE6F5",
            "borderTopWidth": 1,
            "flexDirection": "row",
            "marginTop": 13.5,
            "paddingVertical": 6.5,
            "textAlignVertical": "center",
          }
        }
        viewStyle={
          Object {
            "alignItems": "baseline",
            "borderBottomWidth": 1,
            "borderColor": "#CDE6F5",
            "borderTopWidth": 1,
            "marginTop": 13.5,
            "paddingVertical": 6.5,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "flex-start",
              "flex": 1,
            }
          }
        >
          <View
            style={
              Object {
                "overflow": "hidden",
              }
            }
          >
            <View
              replaceTheme={[Function]}
              updateTheme={[Function]}
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "justifyContent": "center",
                  }
                }
              >
                <Text
                  allowFontScaling={false}
                  selectable={false}
                  style={
                    Array [
                      Object {
                        "color": "#CECABA",
                        "fontSize": 12,
                      },
                      Object {
                        "backgroundColor": "transparent",
                      },
                      Object {
                        "fontFamily": "Feather",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                  testID="iconIcon"
                >
                  
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 3,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#CECABA",
                "fontSize": 16,
              }
            }
          >
            - / - / -
          </Text>
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 3,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#CECABA",
                "fontSize": 16,
              }
            }
          >
            12 / 40 / 20-30
          </Text>
        </View>
        <View
          style={
            Object {
              "alignItems": "flex-end",
              "flex": 1,
            }
          }
        >
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              Array [
                Object {
                  "color": "#CECABA",
                  "fontSize": 15,
                },
                undefined,
                Object {
                  "fontFamily": "Material Design Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            󰇝
          </Text>
        </View>
      </View>
    </View>
    <Modal
      hardwareAccelerated={false}
      onRequestClose={[Function]}
      transparent={true}
      visible={false}
    />
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-around",
        "textAlignVertical": "center",
        "width": "100%",
      }
    }
    viewStyle={
      Object {
        "justifyContent": "space-around",
        "width": "100%",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "borderRadius": 3,
            "overflow": "hidden",
          },
          Object {
            "borderRadius": 3,
          },
          Array [
            Object {
              "height": 53.5,
              "padding": 0,
            },
            Object {
              "alignSelf": "center",
              "marginTop": 10,
              "width": "60%",
            },
          ],
          false,
        ]
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          Object {
            "busy": false,
            "disabled": false,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#31302C",
              "borderColor": "#5F758E",
              "borderRadius": 8,
              "borderWidth": 0,
              "flexDirection": "row",
              "height": 53.5,
              "justifyContent": "center",
              "padding": 8,
              "paddingBottom": 0,
              "paddingTop": 0,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#CECABA",
                "fontSize": 16,
                "paddingVertical": 1,
                "textAlign": "center",
              }
            }
          >
            Add Set
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "borderRadius": 3,
            "overflow": "hidden",
          },
          Object {
            "borderRadius": 3,
          },
          Array [
            Object {
              "height": 53.5,
              "padding": 0,
            },
            Object {
              "alignSelf": "center",
              "marginTop": 10,
              "width": "20%",
            },
          ],
          false,
        ]
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          Object {
            "busy": false,
            "disabled": false,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#31302C",
              "borderColor": "#5F758E",
              "borderRadius": 8,
              "borderWidth": 0,
              "flexDirection": "row",
              "height": 53.5,
              "justifyContent": "center",
              "padding": 8,
              "paddingBottom": 0,
              "paddingTop": 0,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#CECABA",
                "fontSize": 16,
                "paddingVertical": 1,
                "textAlign": "center",
              }
            }
          >
            +WU
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
